crates/daphne-worker/src/storage_proxy/middleware.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2024 Cloudflare, Inc. All rights reserved.
+// SPDX-License-Identifier: BSD-3-Clause
+
+use std::{
+    sync::{Arc, OnceLock},
+    time::Duration,
+};
+
+use axum::{
+    extract::{Path, State},
+    middleware::Next,
+    response::IntoResponse,
+};
+use axum_extra::{
+    headers::{authorization::Bearer, Authorization},
+    TypedHeader,
+};
+use daphne::messages::constant_time_eq;
+use http::{Method, StatusCode};
+use tower_service::Service;
+
+use super::RequestContext;
+
+/// Check if the request's authorization. If unauthorized, return the reason why.
+pub async fn unauthorized_reason(
+    ctx: State<Arc<RequestContext>>,
+    bearer: TypedHeader<Authorization<Bearer>>,
+    request: axum::extract::Request,
+    mut next: Next,
+) -> axum::response::Response {
+    static TRUSTED_TOKEN: OnceLock<Option<String>> = OnceLock::new();
+
+    let Some(trusted_token) = TRUSTED_TOKEN.get_or_init(|| {
+        ctx.env
+            .var("DAPHNE_SERVER_AUTH_TOKEN")
+            .ok()
+            .map(|t| t.to_string())
+    }) else {
+        tracing::warn!("trusted bearer token not configured");
+        return (
+            StatusCode::INTERNAL_SERVER_ERROR,
+            "Authorization token for storage proxy is not configured",
+        )
+            .into_response();
+    };
+
+    if !constant_time_eq(bearer.token().as_bytes(), trusted_token.as_bytes()) {
+        return (StatusCode::UNAUTHORIZED, "Incorrect authorization token").into_response();
+    }
+
+    next.call(request.map(axum::body::Body::new)).await.unwrap()
```

Comment:
Why is panicking safe here?

Reply:
it is safe, but I made it more explicit. This will look better by the end of the year